File: backend/app/storage/local_storage.py
import json
import re
from pathlib import Path

from fastapi import UploadFile

from app.core.exceptions import AppError
from app.models.lesson_document import LessonDocument


class LocalStorage:
    def __init__(self, root: Path, max_file_size_mb: int) -> None:
        self.root = root.resolve()
        self.max_bytes = max_file_size_mb * 1024 * 1024
        self.root.mkdir(parents=True, exist_ok=True)

    def job_dir(self, job_id: str) -> Path:
        if not re.fullmatch(r"[a-f0-9-]{36}", job_id):
            raise AppError("ASSET_NOT_FOUND", "资源不存在", 404)
        path = (self.root / job_id).resolve()
        if not path.is_relative_to(self.root):
            raise AppError("ASSET_NOT_FOUND", "资源不存在", 404)
        return path

    async def save_upload(self, job_id: str, upload: UploadFile) -> Path:
        source_dir = self.job_dir(job_id)
        source_dir.mkdir(parents=True, exist_ok=False)
        suffix = Path(upload.filename or "source").suffix.lower()
        target = source_dir / f"source{suffix}"
        written = 0
        try:
            with target.open("wb") as output:
                while chunk := await upload.read(1024 * 1024):
                    written += len(chunk)
                    if written > self.max_bytes:
                        raise AppError("FILE_TOO_LARGE", "文件大小超过限制", 413)
                    output.write(chunk)
        except Exception:
            target.unlink(missing_ok=True)
            raise
        finally:
            await upload.close()
        return target

    def save_document(self, job_id: str, document: LessonDocument) -> None:
        path = self.job_dir(job_id) / "lesson-document.json"
        path.write_text(document.model_dump_json(by_alias=True, indent=2), encoding="utf-8")

    def save_provider_result(self, job_id: str, result: dict) -> None:
        path = self.job_dir(job_id) / "mineru-result.json"
        serializable = {key: value for key, value in result.items() if key != "result_dir"}
        path.write_text(json.dumps(serializable, ensure_ascii=False, indent=2), encoding="utf-8")

    def resolve_asset(self, job_id: str, filename: str) -> Path:
        assets_dir = (self.job_dir(job_id) / "assets").resolve()
        path = (assets_dir / filename).resolve()
        if not path.is_relative_to(assets_dir) or not path.is_file():
            raise AppError("ASSET_NOT_FOUND", "资源不存在", 404)
        return path

    def resolve_source(self, job_id: str) -> Path:
        job_dir = self.job_dir(job_id)
        matches = sorted(job_dir.glob("source.*"))
        if len(matches) != 1 or not matches[0].is_file():
            raise AppError("SOURCE_NOT_FOUND", "源文件不存在", 404)
        return matches[0]

File: backend-java/src/main/java/com/lessonweb/lesson/storage/LocalStorage.java
package com.lessonweb.lesson.storage;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.lessonweb.lesson.config.LessonProperties;
import com.lessonweb.lesson.exception.AppException;
import com.lessonweb.lesson.parser.MineruParseResult;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Component;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.regex.Pattern;

@Component
public class LocalStorage {

    private static final Pattern JOB_ID_PATTERN = Pattern.compile("[a-f0-9-]{36}");
    private static final int BUFFER_SIZE = 1024 * 1024;

    private final Path root;
    private final long maxBytes;
    private final ObjectMapper objectMapper;

    public LocalStorage(LessonProperties properties, ObjectMapper objectMapper) {
        this.root = Path.of(properties.dataDir()).toAbsolutePath().normalize();
        this.maxBytes = properties.maxFileSize().toBytes();
        this.objectMapper = objectMapper;
        try {
            Files.createDirectories(root);
        } catch (IOException exception) {
            throw new IllegalStateException("无法创建任务数据目录", exception);
        }
    }

    public Path jobDir(String jobId) {
        if (jobId == null || !JOB_ID_PATTERN.matcher(jobId).matches()) {
            throw assetNotFound();
        }
        Path path = root.resolve(jobId).normalize();
        if (!path.startsWith(root)) {
            throw assetNotFound();
        }
        return path;
    }

    public Path saveUpload(String jobId, String filename, MultipartFile upload) {
        Path sourceDir = jobDir(jobId);
        String extension = extensionOf(filename);
        Path target = sourceDir.resolve("source" + extension);
        try {
            Files.createDirectory(sourceDir);
            long written = 0;
            byte[] buffer = new byte[BUFFER_SIZE];
            try (InputStream input = upload.getInputStream(); OutputStream output = Files.newOutputStream(target)) {
                int count;
                while ((count = input.read(buffer)) != -1) {
                    written += count;
                    if (written > maxBytes) {
                        throw new AppException("FILE_TOO_LARGE", "文件大小超过限制", HttpStatus.PAYLOAD_TOO_LARGE);
                    }
                    output.write(buffer, 0, count);
                }
            }
            return target;
        } catch (AppException exception) {
            deleteFailedUpload(sourceDir, target);
            throw exception;
        } catch (IOException exception) {
            deleteFailedUpload(sourceDir, target);
            throw new AppException("UPLOAD_SAVE_FAILED", "上传文件保存失败", HttpStatus.INTERNAL_SERVER_ERROR);
        }
    }

    public void saveProviderResult(String jobId, MineruParseResult result) {
        ObjectNode payload = objectMapper.createObjectNode();
        payload.set("content_list", result.contentList());
        payload.set("ocr_layout", result.ocrLayout());
        writeJson(jobDir(jobId).resolve("mineru-result.json"), payload);
    }

    public Path resolveAsset(String jobId, String filename) {
        Path assetsDir = jobDir(jobId).resolve("assets").normalize();
        Path path = assetsDir.resolve(filename == null ? "" : filename).normalize();
        if (!path.startsWith(assetsDir) || !Files.isRegularFile(path)) {
            throw assetNotFound();
        }
        return path;
    }

    private void writeJson(Path path, Object value) {
        try {
            objectMapper.writerWithDefaultPrettyPrinter().writeValue(path.toFile(), value);
        } catch (IOException exception) {
            throw new AppException("STORAGE_WRITE_FAILED", "任务结果保存失败", HttpStatus.INTERNAL_SERVER_ERROR);
        }
    }

    private String extensionOf(String filename) {
        int dot = filename.lastIndexOf('.');
        return dot < 0 ? "" : filename.substring(dot).toLowerCase();
    }

    private void deleteFailedUpload(Path sourceDir, Path target) {
        try {
            Files.deleteIfExists(target);
            Files.deleteIfExists(sourceDir);
        } catch (IOException ignored) {
            // The original upload error is more useful to the API caller.
        }
    }

    private AppException assetNotFound() {
        return new AppException("ASSET_NOT_FOUND", "资源不存在", HttpStatus.NOT_FOUND);
    }
}
